### ISSUE TITLE
SSL Error when opening the app

### ISSUE THREAD

**Issue by cyteng92** [NYUCCL/psiTurk#297]
I've been using PsiTurk for a while.  Starting from last week, when I try to launch the psiturk app to debug locally, I got the following error message:

Traceback (most recent call last):
  File "/usr/local/bin/psiturk", line 9, in <module>
    load_entry_point('PsiTurk==2.2.1', 'console_scripts', 'psiturk')()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 15, in process
    launch_shell()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 111, in launch_shell
    ps.run(cabinmode=args.cabinmode)
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_shell.py", line 1945, in run
    config.get('psiTurk Access', 'psiturk_secret_access_id'))
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 32, in __init__
    if not self.check_credentials():
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 43, in check_credentials
    auth=(self.access_key, self.secret_key))
  File "/Library/Python/2.7/site-packages/requests/api.py", line 55, in get
    return request('get', url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/api.py", line 44, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 383, in request
    resp = self.send(prep, **send_kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 486, in send
    r = adapter.send(request, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/adapters.py", line 385, in send
    raise SSLError(e)
requests.exceptions.SSLError: EOF occurred in violation of protocol (_ssl.c:590)

-----------------------
I did not change anything in my file and it worked fine last time I used it.  This also happens to a few of my colleagues.  We've all been using it and all of a sudden the above error happens.

I tested it on the school network and wifi at home, and got the same error.  Does anyone have an idea what caused this?  Any comment will be greatly appreciated!

BTW, I'm using psiturk 2.2.1

**Comment by deargle** [comment]
It reaches out to psiturk.org or some url like that on boot. In this part:

if not self.check_credentials():
File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py",
line 43, in check_credentials

This is probably related to the server overload. You can hack it to not
check credentials, if you're not going to use the psiturk ad server.

**Comment by deargle** [comment]
Is it not booting for you right now? psiturk starts fine for me at the moment.

**Comment by gureckis** [comment]
i doubt it is related to the server overload.  the ad server and the server that is touched on the boot are distinct and only the ad server is likely to ever have significant load.  is looks to me like libraries related to SSL were updated on the local install and there is some incompatibility.  Please check stackoverflow for more info about that error message.  A quick check myself shows that you may need to check you have the proper SSL libraries installed in python: https://stackoverflow.com/questions/33410577/python-requests-exceptions-sslerror-eof-occurred-in-violation-of-protocol

**Comment by cyteng92** [comment]
Thanks for the quick reply!  It is still not working locally on my laptop (the cabin mode works).  I'll try to look into the SLL libraries and let you know how it turns out.  Thanks!

**Comment by cyteng92** [comment]
I tried to install the SSL libraries and it seems that I don't have ndg-httpsclient installed.
After the installation, it says:

Traceback (most recent call last):
  File "/usr/local/bin/psiturk", line 9, in <module>
    load_entry_point('PsiTurk==2.2.1', 'console_scripts', 'psiturk')()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 15, in process
    launch_shell()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 111, in launch_shell
    ps.run(cabinmode=args.cabinmode)
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_shell.py", line 1945, in run
    config.get('psiTurk Access', 'psiturk_secret_access_id'))
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 32, in __init__
    if not self.check_credentials():
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 43, in check_credentials
    auth=(self.access_key, self.secret_key))
  File "/Library/Python/2.7/site-packages/requests/api.py", line 55, in get
    return request('get', url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/api.py", line 44, in request
    return session.request(method=method, url=url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 383, in request
    resp = self.send(prep, **send_kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 486, in send
    r = adapter.send(request, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/adapters.py", line 385, in send
    raise SSLError(e)
requests.exceptions.SSLError

**Comment by cyteng92** [comment]
I'm not planning to use psiturk ad server.  Is there a way to make it to not check credentials?  Thanks

**Comment by gureckis** [comment]
the issue you are having is with the `requests` library and possibly your mac os x SSL environment and doesn't have anything to do specifically with psiturk per-se.  but certainly you can try to remove the psiturk dependence on the `requests` library in order to use the rest of the code.

**Comment by deargle** [comment]
Try this: https://stackoverflow.com/a/39580231

Failing fixing your ssl issues (I recommend you keep trying to fix them), you can hack on psiturk by doing this (I hope you're using a virtualenv):

```
pip uninstall psiturk
git clone https://github.com/NYUCCL/psiTurk.git
pip install -e psiTurk
```

Then edit the files in the psiTurk folder

Work around here: 

File "/Library/Python/2.7/site-packages/**psiturk/psiturk_org_services.py**", **line 32**, in init
if not self.check_credentials():

**Comment by deargle** [comment]
If you don't fix ssl, you'll probably have other things break, like hit creation, when it reaches out to the mturk servers via ssl

**Comment by cyteng92** [comment]
Thanks a lot for all the suggestions. I tried to install the SSL libraries and updated psiturk to the newest version.  This time I got a different error when starting psiturk:

Traceback (most recent call last):
  File "/usr/local/bin/psiturk", line 9, in <module>
    load_entry_point('PsiTurk==2.2.3', 'console_scripts', 'psiturk')()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 15, in process
    launch_shell()
  File "/Library/Python/2.7/site-packages/psiturk/command_line.py", line 111, in launch_shell
    ps.run(cabinmode=args.cabinmode)
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_shell.py", line 1946, in run
    config.get('psiTurk Access', 'psiturk_secret_access_id'))
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 32, in __init__
    if not self.check_credentials():
  File "/Library/Python/2.7/site-packages/psiturk/psiturk_org_services.py", line 43, in check_credentials
    auth=(self.access_key, self.secret_key))
  File "/Library/Python/2.7/site-packages/requests/api.py", line 65, in get
    return request('get', url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/api.py", line 49, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 461, in request
    resp = self.send(prep, **send_kwargs)
  File "/Library/Python/2.7/site-packages/requests/sessions.py", line 573, in send
    r = adapter.send(request, **kwargs)
  File "/Library/Python/2.7/site-packages/requests/adapters.py", line 370, in send
    timeout=timeout
  File "/Library/Python/2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 518, in urlopen
    body=body, headers=headers)
  File "/Library/Python/2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 325, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=conn.timeout)
  File "/Library/Python/2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 295, in _raise_timeout
    if 'timed out' in str(err) or 'did not complete (read)' in str(err):  # Python 2.6
TypeError: __str__ returned non-string (type SysCallError)

-----------
When I tried to set the verify parameter, it sends back the same error message.

**Comment by deargle** [comment]
Might not help, but sometimes using a virtualenv gives you a fresh start
and can sidestep some of these issues. You can tinker with different
`requests` versions without messing up the system for other users.

**Comment by cyteng92** [comment]
Thank you! I'll look into this.

**Comment by deargle** [comment]
Someone please report back about whether a virtualenv fixes the problem

**Comment by cyteng92** [comment]
I installed virtualenv and launched psiturk within it.  Still getting the same error though..

The next thing I tried was to hack the check_credential function and set verify= False, then when I launched the psiturk app, it just hanged for a long time and never started.

After I disabled the check_credentials function, I started to get the error message from the urllib2 library, instead of requests.  Here is the error message:

Traceback (most recent call last):
  File "/Users/cteng/ENV/bin/psiturk", line 11, in <module>
    sys.exit(process())
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/command_line.py", line 15, in process
    launch_shell()
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/command_line.py", line 111, in launch_shell
    ps.run(cabinmode=args.cabinmode)
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/psiturk_shell.py", line 1949, in run
    config.getboolean('Shell Parameters', 'launch_in_sandbox_mode'), quiet=quiet)
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/psiturk_shell.py", line 550, in __init__
    PsiturkShell.__init__(self, config, server)
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/psiturk_shell.py", line 124, in __init__
    self.intro = self.get_intro_prompt()
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/psiturk_shell.py", line 605, in get_intro_prompt
    server_msg = self.web_services.get_system_status()
  File "/Users/cteng/ENV/lib/python2.7/site-packages/psiturk/psiturk_org_services.py", line 73, in get_system_status
    response = urllib2.urlopen(api_server_status_link, timeout=1)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 431, in open
    response = self._open(req, data)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 449, in _open
    '_open', req)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1240, in https_open
    context=self._context)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 1197, in do_open
    raise URLError(err)
urllib2.URLError: <urlopen error EOF occurred in violation of protocol (_ssl.c:590)>

**Comment by deargle** [comment]
Try installing a different python:
https://github.com/mcs07/CIRpy/issues/3#issuecomment-171820410

**Comment by cyteng92** [comment]
I am currently using python 2.7.10. What version would you recommend to use? Should I bump it down to 2.6, or try a newer version? Thanks

**Comment by deargle** [comment]
No, I'd stay on 2.7, but just try following that link to install python 2.7
from brew and see what happens. It's a different distribution. They say
there's something wonky about the mac python default install. I'm not a mac
user so can't confirm.

This link: https://github.com/mcs07/CIRpy/issues/3#issuecomment-171820410

**Comment by lilyz** [comment]
I had the same issue. I fixed it just now by using the version of python installed using brew.

**Comment by cyteng92** [comment]
Just a little update, I fixed it by installing python 2.7.14.  

It is still a bit strange to me that pisturk worked fine with python 2.7.10 previously but became incompatible since Feb 2nd.  It might be worthwhile to make a note about the python version in the documentation.

Thank you all for the help!
